sips/sip-010/sip-010-fungible-token-standard.md

```diff
@@ -0,0 +1,132 @@
+
+## Specification
+
+The fungible token trait, `ft-trait`, has a few methods:
+
+### Transfer
+
+`(transfer ((recipient principal) (amount uint)) (response bool uint))`
+
+Transfer the fungible token from the sender of this transaction to the recipient. The `amount` is an unsigned integer. It is recommended that implementing contracts use the built-in `ft-transfer` Clarity method. If the sender does not have enough tokens to complete the transaction, the transaction should abort and return an `(err uint)`.
+
+This method must be defined with `define-public`, as it alters state, and should be externally callable.
+
+### Name
+
+`(name () (response (string-ascii 32) uint))`
+
+Return a human-readable name for the contract, such as "CoolPoints", etc.
+
+This method should be defined as read-only, i.e. `define-read-only`.
+
+### Symbol
+
+`(symbol () (response (string-ascii 32) uint))`
+
+Return a symbol that allows for a shorter representation of your token. This is sometimes referred to as a "ticker". Examples: "STX", "COOL", etc. Typically, your token could be referred to as $SYMBOL when referencing it in writing.
+
+This method should be defined as read-only, i.e. `define-read-only`.
+
+### Decimals
+
+`(decimals () (response uint uint))`
+
+The number of decimal places in your token. All fungible token balances must be represented as integers, but providing the number of decimals provides for an abstraction of your token that humans are more familiar dealing with. For example, the US Dollar has 2 decimals, if the base unit is "cents", as is typically done in accounting. Stacks has 6 decimals, Bitcoin has 8 decimals, and so on.
+
+As another example, if your token has 4 decimals, and the `balance-of` a particular user returns `100345000`, wallets and exchanges would likely represent that value as `10034.5`.
+
+This method should be defined as read-only, i.e. `define-read-only`.
+
+### Balance of
+
+`(balance-of (principal) (response uint uint))`
+
+Return the balance of a particular principal (also known as "address" or "account"). Implementations should typically use the built-in Clarity method `ft-get-balance`.
+
+This method should be defined as read-only, i.e. `define-read-only`.
+
+### Total supply
+
+`(total-supply () (response uint uint))`
+
+Return the total supply of this token. Implementations should typically use the built-in Clarity method `ft-get-supply`.
+
+This method should be defined as read-only, i.e. `define-read-only`.
+
+## Trait implementation
+
+An implementation of the proposed trait is provided below.
+
+```clarity
+(define-trait ft-trait
+  (
+    ;; Transfer from the caller to a new principal
+    (transfer (principal uint) (response bool uint))
+
+    ;; the human readable name of the token
+    (name () (response (string-ascii 32) uint))
+
+    ;; the ticker symbol, or empty if none
```

Comment:
empty string is easier to deal with

Reply:
I had chosen that order to make it easier for Solidity refugees to get their foot in the door... 😉

Reply:
but then they have to learn that the trait is different to the native function.

Reply:
Thoughts @friedger @psq? IMO a blank string is easier to deal with on clients than an `optional`.

Reply:
> IMO a blank string is easier to deal with on clients than an optional. (edited to remove mentions - my Github hadn't updated when I commented).

I'm curious how Web clients will be interacting with this trait?  If they're going to be using `stacks.js` most of the time, then maybe it would make more sense to have `(response (optional (string-ascii 32)) uint)`, since the library will already take care of parsing the Clarity return value into a usable struct?

Reply:
Yeah, stacks.js already has solid support for `deserializeCV(optionalAsBuffer: Buffer)`. But, does it ever make sense to not have a token name?

If we use an optional, there is no way to guarantee that the value of the optional is not a blank string. So, then clients basically have to handle two special cases - a blank string, or a `none`, as compared to just handling a blank string.

Reply:
I think it's ultimately up to the SIP authors to choose what the appropriate action should be :)  I'm only trying to make sure that alternative options are considered.

I think you're probably right that `(some "")` and `none` would end up representing the same thing (a nameless token), so it wouldn't make sense to have two different representations of the same concept.  So, maybe a better set of questions would be:

* Must a token have a name?
* What would it mean for a token to _not_ have a name?
* If a token has a name, must the token's name be globally unique, and if so, how might this be enforced?

Reply:
Moving this along, where did we land on this?  How about the following?

* A token can have any ASCII-encoded name it wants, up to 32 bytes.  The return type here is appropriate -- `(response (string-ascii 32) uint)`.
* A token can have a zero-byte name.
* A token will report exactly one name through the `name` method.
* Other people can make their own indexes over tokens.  Downstream dependents are advised to consider the token's self-reported name and symbol as hints, because there is no enforcement mechanism defined by this SIP that would ensure that names correspond to the token implementation that the downstream dependent expects.
* Token names and symbols do not have to be globally unique.  They are hints which downstream dependents can ignore.

Reply:
Haven't heard anything back on this, and I think what I've suggested here is consistent with the SIP.  Going to go ahead and resolve this comment thread.